twisted/names/dns.py

```diff
@@ -2073,16 +2073,25 @@ def _compactRepr(obj, alwaysShow=None, flagNames=None, fieldNames=None,
 
     out = ['<', obj.__class__.__name__]
 
-    # Get the argument names and values from the constructor.
-    argspec = inspect.getargspec(obj.__class__.__init__)
-    # Reverse the args and defaults to avoid mapping positional arguments
-    # which don't have a default.
-    defaults = dict(zip(reversed(argspec.args), reversed(argspec.defaults)))
-    for name in fieldNames:
-        defaultValue = defaults.get(name)
-        fieldValue = getattr(obj, name, defaultValue)
-        if (name in alwaysShow) or (fieldValue != defaultValue):
-            out.append(' %s=%r' % (name, fieldValue))
+    if _PY3:
+        # Get the argument names and values from the constructor.
+        signature = inspect.signature(obj.__class__.__init__)
+        for name in fieldNames:
+            defaultValue = signature.parameters[name].default
+            fieldValue = getattr(obj, name, defaultValue)
+            if (name in alwaysShow) or (fieldValue != defaultValue):
+                out.append(' %s=%r' %(name, fieldValue))
+    else:
+        # Get the argument names and values from the constructor.
+        argspec = inspect.getargspec(obj.__class__.__init__)
+        # Reverse the args and defaults to avoid mapping positional arguments
+        # which don't have a default.
+        defaults = dict(zip(reversed(argspec.args), reversed(argspec.defaults)))
+        for name in fieldNames:
+            defaultValue = defaults.get(name)
+            fieldValue = getattr(obj, name, defaultValue)
+            if (name in alwaysShow) or (fieldValue != defaultValue):
+                out.append(' %s=%r' % (name, fieldValue))
```

Comment:
So I appreciate this is the kind of feedback you don't really want for changes like this, but this code block is sufficiently large now that I think it's obscuring the intent of the function. Do you mind hoisting this out to a helper function?